asm/inline.rs

```diff
@@ -20,45 +22,58 @@ pub unsafe fn __control_r() -> u32 {
 
 #[inline(always)]
 pub unsafe fn __control_w(w: u32) {
-    asm!("msr CONTROL, {}", in(reg) w);
+    // ISB is required after writing to CONTROL,
+    // per ARM architectural requirements (see Application Note 321).
+    asm!("msr CONTROL, {}", "isb", in(reg) w);
+
+    // Ensure instructions are not reordered around the CONTROL update.
+    compiler_fence(Ordering::SeqCst);
 }
 
 #[inline(always)]
 pub unsafe fn __cpsid() {
     asm!("cpsid i");
+
+    // Ensure no subsequent instructions are reordered to before interrupts are disabled.
+    compiler_fence(Ordering::SeqCst);
 }
 
 #[inline(always)]
 pub unsafe fn __cpsie() {
+    // Ensure no preceeding instructions are reordered to after interrupts are enabled.
+    compiler_fence(Ordering::SeqCst);
+
     asm!("cpsie i");
 }
 
 #[inline(always)]
 pub unsafe fn __delay(cyc: u32) {
-    asm!("
-    1:
-        nop
-        subs {}, #1
-        bne 1b
-        // Branch to 1 instead of delay does not generate R_ARM_THM_JUMP8 relocation, which breaks
-        // linking on the thumbv6m-none-eabi target
-    ", in(reg) cyc);
+    // Use local labels to avoid R_ARM_THM_JUMP8 relocations which fail on thumbv6m.
+    asm!(
+        "1:",
+        "nop",
+        "subs {}, #1",
+        "bne 1b",
+        in(reg) cyc
+    );
 }
 
-// FIXME do we need compiler fences here or should we expect them in the caller?
 #[inline(always)]
 pub unsafe fn __dmb() {
-    asm!("dmb 0xF");
+    asm!("dmb");
+    compiler_fence(Ordering::AcqRel);
```

Comment:
> It ensures that all explicit memory accesses that appear in program order before the DMB instruction are observed before any explicit memory accesses that appear in program order after the DMB instruction.

Shouldn't this be `SeqCst` ? Consider the following case:

```
read - 1
-> 3 and 4 can be moved to here
write - 2

dmb
AcqRel

read - 3
-> 1 and 2 can be moved to here
write - 4
```

Is my reasoning correct ?

Reply:
Fair, it stood out a bit as the only AcqRel anyway. I've updated it to SeqCst for now.